File: Java/src/main/java/com/DSA/Java/Algorithms/DynamicProgramming/ConvertStringWithDp.java
package com.DSA.Java.Algorithms.DynamicProgramming;

import java.util.HashMap;

public class ConvertStringWithDp {
  private HashMap<String, Integer> memo = new HashMap<>();

  private int findEditsTopDownHelper(String s1, String s2, int index1, int index2) {
    if (index1 == s1.length()) {
      return s2.length() - index2;
    }

    if (index2 == s2.length()) {
      return s1.length() - index1;
    }
    
    if (s1.charAt(index1) == s2.charAt(index2)) {
      return findEditsTopDownHelper(s1, s2, index1 + 1, index2 + 1);
    }

    String dictKey = Integer.toString(index1) + Integer.toString(index2);
    if (!memo.containsKey(dictKey)) {
      int deleteOp = 1 + findEditsTopDownHelper(s1, s2, index1 + 1, index2);
      int insertOp = 1 + findEditsTopDownHelper(s1, s2, index1, index2 + 1);
      int replaceOp = 1 + findEditsTopDownHelper(s1, s2, index1 + 1, index2 + 1);
      memo.put(dictKey, Math.min(deleteOp, Math.min(insertOp, replaceOp)));
    }
    return memo.get(dictKey);
  }

  public int findEditsTopDown(String s1, String s2) {
    return findEditsTopDownHelper(s1, s2, 0, 0);
  }
}

File: C++/Algorithm/src/DynamicProgramming/ConvertStringWithDP.cpp
#include "../../include/DynamicProgramming/ConvertStringWithDP.h"

int ConvertStringWithDP::convert_string_using_top_down_method_helper(const std::string &s1,
                                                                     const std::string &s2,
                                                                     int index_1,
                                                                     int index_2) {
  if (s1.length() == index_1) {
    return s2.length() - index_2;
  }

  if (s2.length() == index_2) {
    return s1.length() - index_1;
  }

  if (s1[index_1] == s2[index_2]) {
    return convert_string_using_top_down_method_helper(s1, s2, index_1 + 1, index_2 + 1);
  }

  std::string map_key = std::to_string(index_1) + std::to_string(index_2);
  if (keys.find(map_key) == keys.end()) {
    int delete_operation = 1 + convert_string_using_top_down_method_helper(s1, s2, index_1 + 1, index_2);
    int replace_operation = 1 + convert_string_using_top_down_method_helper(s1, s2, index_1 + 1, index_2 + 1);
    int insert_operation = 1 + convert_string_using_top_down_method_helper(s1, s2, index_1, index_2 + 1);
    keys[map_key] = std::min(delete_operation, std::min(replace_operation, insert_operation));
  }
  return keys[map_key];
}

int ConvertStringWithDP::convert_string_using_top_down_method(const std::string &s1,
                                                              const std::string &s2) {
  return convert_string_using_top_down_method_helper(s1, s2, 0, 0);
}
